supernode/node/grpc/server/services/walletnode/external_dupe_detection.go

```diff
@@ -0,0 +1,365 @@
+package walletnode
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/base64"
+	"io"
+
+	"github.com/pastelnetwork/gonode/common/errors"
+	"github.com/pastelnetwork/gonode/common/log"
+	pb "github.com/pastelnetwork/gonode/proto/walletnode"
+	"github.com/pastelnetwork/gonode/supernode/node/grpc/server/services/common"
+	"github.com/pastelnetwork/gonode/supernode/services/externaldupedetection"
+	"golang.org/x/crypto/sha3"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+// ExternalDupeDetection represents grpc service for registration artwork.
+type ExternalDupeDetection struct {
+	pb.UnimplementedExternalDupeDetectionServer
+
+	*common.ExternalDupeDetection
+}
+
+// Session implements walletnode.ExternalDupeDetectionServer.Session()
+func (service *ExternalDupeDetection) Session(stream pb.ExternalDupeDetection_SessionServer) error {
+	ctx, cancel := context.WithCancel(stream.Context())
+	defer cancel()
+
+	var task *externaldupedetection.Task
+
+	if sessID, ok := service.SessID(ctx); ok {
+		if task = service.Task(sessID); task == nil {
+			return errors.Errorf("not found %q task", sessID)
+		}
+	} else {
+		task = service.NewTask()
+	}
+	go func() {
+		<-task.Done()
+		cancel()
+	}()
+	defer task.Cancel()
```

Comment:
should cancel task only if it's created new in request.
See bug fix at https://github.com/pastelnetwork/gonode/pull/215